## Table of Contents

- Importing Libraries
- Importing Dataframes
- Exploring Dataframes
- Wrangling Data
- Consistency Checks
- Exporting Dataframes

## Importing Libraries

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

In [2]:
# Define path
path = r'/Users/isaacotubanjo/Documents/US Gun Violence Analysis'

In [3]:
path

'/Users/isaacotubanjo/Documents/US Gun Violence Analysis'

## Importing Dataframes

In [4]:
# Importing dataframe using os path
df_gun = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'gun-violence-data.csv'), index_col = False)

## Exploring Dataframes

In [5]:
# Finding the names of columns in the dataframe
df_gun.columns

Index(['incident_id', 'date', 'state', 'city_or_county', 'address', 'n_killed',
       'n_injured', 'incident_url', 'source_url',
       'incident_url_fields_missing', 'congressional_district', 'gun_stolen',
       'gun_type', 'incident_characteristics', 'latitude',
       'location_description', 'longitude', 'n_guns_involved', 'notes',
       'participant_age', 'participant_age_group', 'participant_gender',
       'participant_name', 'participant_relationship', 'participant_status',
       'participant_type', 'sources', 'state_house_district',
       'state_senate_district'],
      dtype='object')

In [6]:
# Checking the first 20 rows of the dataframe
df_gun.head(20)

incident_id        date                 state city_or_county  \
0        461105  2013-01-01          Pennsylvania     Mckeesport   
1        460726  2013-01-01            California      Hawthorne   
2        478855  2013-01-01                  Ohio         Lorain   
3        478925  2013-01-05              Colorado         Aurora   
4        478959  2013-01-07        North Carolina     Greensboro   
5        478948  2013-01-07              Oklahoma          Tulsa   
6        479363  2013-01-19            New Mexico    Albuquerque   
7        479374  2013-01-21             Louisiana    New Orleans   
8        479389  2013-01-21            California      Brentwood   
9        492151  2013-01-23              Maryland      Baltimore   
10       491674  2013-01-23             Tennessee    Chattanooga   
11       479413  2013-01-25              Missouri    Saint Louis   
12       479561  2013-01-26             Louisiana      Charenton   
13       479554  2013-01-26  District of Columbia     Washington   
14       479460  2013-01-26                  Ohio    Springfield   
15       479573  2013-02-02             Tennessee        Memphis   
16       479580  2013-02-03            California  Yuba (county)   
17       479592  2013-02-07              Illinois        Chicago   
18       479603  2013-02-09             Louisiana    New Orleans   
19       480311  2013-02-11            California        Vallejo   

                                              address  n_killed  n_injured  \
0           1506 Versailles Avenue and Coursin Street         0          4   
1                        13500 block of Cerise Avenue         1          3   
2                               1776 East 28th Street         1          3   
3                    16000 block of East Ithaca Place         4          0   
4                           307 Mourning Dove Terrace         2          2   
5                          6000 block of South Owasso         4          0   
6                                      2806 Long Lane         5          0   
7   LaSalle Street and Martin Luther King Jr. Boul...         0          5   
8                          1100 block of Breton Drive         0          4   
9                        1500 block of W. Fayette St.         1          6   
10                                     1501 Dodds Ave         1          3   
11                W Florissant Ave and Riverview Blvd         1          3   
12                       1000 block of Flat Town Road         2          3   
13                        2403 Benning Road Northeast         0          5   
14                               601 West Main Street         1          3   
15                                  2514 Mount Moriah         0          5   
16                        5800 block of Poplar Avenue         1          3   
17                     2500 block of East 75th Street         0          4   
18                        400 block of Bourbon Street         0          4   
19                       800 block of Humboldt Street         1          4   

                                         incident_url  \
0   http://www.gunviolencearchive.org/incident/461105   
1   http://www.gunviolencearchive.org/incident/460726   
2   http://www.gunviolencearchive.org/incident/478855   
3   http://www.gunviolencearchive.org/incident/478925   
4   http://www.gunviolencearchive.org/incident/478959   
5   http://www.gunviolencearchive.org/incident/478948   
6   http://www.gunviolencearchive.org/incident/479363   
7   http://www.gunviolencearchive.org/incident/479374   
8   http://www.gunviolencearchive.org/incident/479389   
9   http://www.gunviolencearchive.org/incident/492151   
10  http://www.gunviolencearchive.org/incident/491674   
11  http://www.gunviolencearchive.org/incident/479413   
12  http://www.gunviolencearchive.org/incident/479561   
13  http://www.gunviolencearchive.org/incident/479554   
14  http://www.gunviolencearchive.org/incident/479460   
15  http://www.gunviolen

In [7]:
df_gun.shape

(239677, 29)

In [8]:
# Checking the descriptive statistics for the dataframe
df_gun.describe()

incident_id       n_killed      n_injured  congressional_district  \
count  2.396770e+05  239677.000000  239677.000000           227733.000000   
mean   5.593343e+05       0.252290       0.494007                8.001265   
std    2.931287e+05       0.521779       0.729952                8.480835   
min    9.211400e+04       0.000000       0.000000                0.000000   
25%    3.085450e+05       0.000000       0.000000                2.000000   
50%    5.435870e+05       0.000000       0.000000                5.000000   
75%    8.172280e+05       0.000000       1.000000               10.000000   
max    1.083472e+06      50.000000      53.000000               53.000000   

            latitude      longitude  n_guns_involved  state_house_district  \
count  231754.000000  231754.000000    140226.000000         200905.000000   
mean       37.546598     -89.338348         1.372442             55.447132   
std         5.130763      14.359546         4.678202             42.048117   
min        19.111400    -171.429000         1.000000              1.000000   
25%        33.903400     -94.158725         1.000000             21.000000   
50%        38.570600     -86.249600         1.000000             47.000000   
75%        41.437375     -80.048625         1.000000             84.000000   
max        71.336800      97.433100       400.000000            901.000000   

       state_senate_district  
count           207342.00000  
mean                20.47711  
std                 14.20456  
min                  1.00000  
25%                  9.00000  
50%                 19.00000  
75%                 30.00000  
max                 94.00000

There might be outliers in the n_killed, n_injured and n_guns_involved columns. This is apparent when you compare the mean and the max values.

In [9]:
# Checking the datatype for columns
df_gun.dtypes

incident_id                      int64
date                            object
state                           object
city_or_county                  object
address                         object
n_killed                         int64
n_injured                        int64
incident_url                    object
source_url                      object
incident_url_fields_missing       bool
congressional_district         float64
gun_stolen                      object
gun_type                        object
incident_characteristics        object
latitude                       float64
location_description            object
longitude                      float64
n_guns_involved                float64
notes                           object
participant_age                 object
participant_age_group           object
participant_gender              object
participant_name                object
participant_relationship        object
participant_status              object
participant_type         

## Wrangling Data

In [10]:
# Some columns need to be dropped as they are irrelevant to the analysis
df_gun = df_gun.drop(columns = ['incident_url', 'source_url', 'incident_url_fields_missing', 'participant_name', 'participant_relationship', 'notes', 'sources'])

In [11]:
# Checking results
df_gun.head(20)

incident_id        date                 state city_or_county  \
0        461105  2013-01-01          Pennsylvania     Mckeesport   
1        460726  2013-01-01            California      Hawthorne   
2        478855  2013-01-01                  Ohio         Lorain   
3        478925  2013-01-05              Colorado         Aurora   
4        478959  2013-01-07        North Carolina     Greensboro   
5        478948  2013-01-07              Oklahoma          Tulsa   
6        479363  2013-01-19            New Mexico    Albuquerque   
7        479374  2013-01-21             Louisiana    New Orleans   
8        479389  2013-01-21            California      Brentwood   
9        492151  2013-01-23              Maryland      Baltimore   
10       491674  2013-01-23             Tennessee    Chattanooga   
11       479413  2013-01-25              Missouri    Saint Louis   
12       479561  2013-01-26             Louisiana      Charenton   
13       479554  2013-01-26  District of Columbia     Washington   
14       479460  2013-01-26                  Ohio    Springfield   
15       479573  2013-02-02             Tennessee        Memphis   
16       479580  2013-02-03            California  Yuba (county)   
17       479592  2013-02-07              Illinois        Chicago   
18       479603  2013-02-09             Louisiana    New Orleans   
19       480311  2013-02-11            California        Vallejo   

                                              address  n_killed  n_injured  \
0           1506 Versailles Avenue and Coursin Street         0          4   
1                        13500 block of Cerise Avenue         1          3   
2                               1776 East 28th Street         1          3   
3                    16000 block of East Ithaca Place         4          0   
4                           307 Mourning Dove Terrace         2          2   
5                          6000 block of South Owasso         4          0   
6                                      2806 Long Lane         5          0   
7   LaSalle Street and Martin Luther King Jr. Boul...         0          5   
8                          1100 block of Breton Drive         0          4   
9                        1500 block of W. Fayette St.         1          6   
10                                     1501 Dodds Ave         1          3   
11                W Florissant Ave and Riverview Blvd         1          3   
12                       1000 block of Flat Town Road         2          3   
13                        2403 Benning Road Northeast         0          5   
14                               601 West Main Street         1          3   
15                                  2514 Mount Moriah         0          5   
16                        5800 block of Poplar Avenue         1          3   
17                     2500 block of East 75th Street         0          4   
18                        400 block of Bourbon Street         0          4   
19                       800 block of Humboldt Street         1          4   

    congressional_district              gun_stolen  \
0                     14.0                     NaN   
1                     43.0                     NaN   
2                      9.0  0::Unknown||1::Unknown   
3                      6.0                     NaN   
4                      6.0  0::Unknown||1::Unknown   
5                      1.0                     NaN   
6                      1.0  0::Unknown||1::Unknown   
7                      2.0                     NaN   
8                      9.0                     NaN   
9                      7.0                     NaN   
10                     3.0              0::Unknown   
11                     1.0              0::Unknown   
12                     3.0              0::Unknown   
13                     1.0              0::Unknown   
14                     8.0                     NaN   
15                     9.0              0::Unknown   
16                

In [12]:
df_gun.shape

(239677, 22)

In [13]:
# Changing incident_id to string as it is unnecessary for analysis
df_gun ['incident_id'] = df_gun['incident_id'].astype('str')

In [14]:
# Checking results
df_gun['incident_id'].dtype

dtype('O')

In [16]:
# Renaming columns
df_gun.rename(columns = {'n_killed':'number_killed', 'n_injured':'number_injured', 'n_guns_involved':'number_guns_involved'}, inplace = True)

In [17]:
# Checking results
df_gun.head()

incident_id        date           state city_or_county  \
0      461105  2013-01-01    Pennsylvania     Mckeesport   
1      460726  2013-01-01      California      Hawthorne   
2      478855  2013-01-01            Ohio         Lorain   
3      478925  2013-01-05        Colorado         Aurora   
4      478959  2013-01-07  North Carolina     Greensboro   

                                     address  number_killed  number_injured  \
0  1506 Versailles Avenue and Coursin Street              0               4   
1               13500 block of Cerise Avenue              1               3   
2                      1776 East 28th Street              1               3   
3           16000 block of East Ithaca Place              4               0   
4                  307 Mourning Dove Terrace              2               2   

   congressional_district              gun_stolen                gun_type  \
0                    14.0                     NaN                     NaN   
1                    43.0                     NaN                     NaN   
2                     9.0  0::Unknown||1::Unknown  0::Unknown||1::Unknown   
3                     6.0                     NaN                     NaN   
4                     6.0  0::Unknown||1::Unknown  0::Handgun||1::Handgun   

   ... location_description  longitude number_guns_involved  \
0  ...                  NaN   -79.8559                  NaN   
1  ...                  NaN  -118.3330                  NaN   
2  ...          Cotton Club   -82.1377                  2.0   
3  ...                  NaN  -104.8020                  NaN   
4  ...                  NaN   -79.9569                  2.0   

                     participant_age  \
0                              0::20   
1                              0::20   
2  0::25||1::31||2::33||3::34||4::33   
3         0::29||1::33||2::56||3::33   
4         0::18||1::46||2::14||3::47   

                               participant_age_group  \
0  0::Adult 18+||1::Adult 18+||2::Adult 18+||3::A...   
1  0::Adult 18+||1::Adult 18+||2::Adult 18+||3::A...   
2  0::Adult 18+||1::Adult 18+||2::Adult 18+||3::A...   
3  0::Adult 18+||1::Adult 18+||2::Adult 18+||3::A...   
4  0::Adult 18+||1::Adult 18+||2::Teen 12-17||3::...   

                            participant_gender  \
0         0::Male||1::Male||3::Male||4::Female   
1                                      0::Male   
2  0::Male||1::Male||2::Male||3::Male||4::Male   
3         0::Female||1::Male||2::Male||3::Male   
4       0::Female||1::Male||2::Male||3::Female   

                                  participant_status  \
0  0::Arrested||1::Injured||2::Injured||3::Injure...   
1      0::Killed||1::Injured||2::Injured||3::Injured   
2  0::Injured, Unharmed, Arrested||1::Unharmed, A...   
3         0::Killed||1::Killed||2::Killed||3::Killed   
4       0::Injured||1::Injured||2::Killed||3::Killed   

                                    participant_type state_house_district  \
0  0::Victim||1::Victim||2::Victim||3::Victim||4:...                  NaN   
1  0::Victim||1::Victim||2::Victim||3::Victim||4:...                 62.0   
2  0::Subject-Suspect||1::Subject-Suspect||2::Vic...                 56.0   
3  0::Victim||1::Victim||2::Victim||3::Subject-Su...                 40.0   
4  0::Victim||1::Victim||2::Victim||3::Subject-Su...                 62.0   

  state_senate_district  
0                   NaN  
1                  35.0  
2                  13.0  
3                  28.0  
4                  27.0  

[5 rows x 22 columns]

## Consistency Checks

In [19]:
# Checking for Mixed Types
for col in df_gun.columns.tolist():
  weird = (df_gun[[col]].applymap(type) != df_gun[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_gun[weird]) > 0:
    print (col)

address
gun_stolen
gun_type
incident_characteristics
location_description
participant_age
participant_age_group
participant_gender
participant_status
participant_type


In [20]:
# Checking for Missing Values
df_gun.isnull().sum()

incident_id                      0
date                             0
state                            0
city_or_county                   0
address                      16497
number_killed                    0
number_injured                   0
congressional_district       11944
gun_stolen                   99498
gun_type                     99451
incident_characteristics       326
latitude                      7923
location_description        197588
longitude                     7923
number_guns_involved         99451
participant_age              92298
participant_age_group        42119
participant_gender           36362
participant_status           27626
participant_type             24863
state_house_district         38772
state_senate_district        32335
dtype: int64

In [21]:
# Checking for Duplicate Values
df_dups = df_gun[df_gun.duplicated()]

In [22]:
df_dups

Empty DataFrame
Columns: [incident_id, date, state, city_or_county, address, number_killed, number_injured, congressional_district, gun_stolen, gun_type, incident_characteristics, latitude, location_description, longitude, number_guns_involved, participant_age, participant_age_group, participant_gender, participant_status, participant_type, state_house_district, state_senate_district]
Index: []

[0 rows x 22 columns]

There are no duplicates in the dataframe

There are a lot of missing values in different columns. We can not remove the rows with missing values as they are relevant to our analysis. The incident_characteristics column has just 326 row with missing values. These can be removed as it is not a substantial amount.

In [23]:
df_gun['incident_characteristics'].head(10)

0    Shot - Wounded/Injured||Mass Shooting (4+ vict...
1    Shot - Wounded/Injured||Shot - Dead (murder, a...
2    Shot - Wounded/Injured||Shot - Dead (murder, a...
3    Shot - Dead (murder, accidental, suicide)||Off...
4    Shot - Wounded/Injured||Shot - Dead (murder, a...
5    Shot - Dead (murder, accidental, suicide)||Hom...
6    Shot - Dead (murder, accidental, suicide)||Mas...
7    Shot - Wounded/Injured||Drive-by (car to stree...
8    Shot - Wounded/Injured||Drive-by (car to stree...
9    Shot - Wounded/Injured||Shot - Dead (murder, a...
Name: incident_characteristics, dtype: object

In [24]:
df_gun_clean = df_gun[df_gun['incident_characteristics'].isnull()  == False]

In [25]:
# Checking results
df_gun_clean.isnull().sum()

incident_id                      0
date                             0
state                            0
city_or_county                   0
address                      16476
number_killed                    0
number_injured                   0
congressional_district       11940
gun_stolen                   99325
gun_type                     99278
incident_characteristics         0
latitude                      7919
location_description        197303
longitude                     7919
number_guns_involved         99278
participant_age              92134
participant_age_group        42030
participant_gender           36294
participant_status           27559
participant_type             24814
state_house_district         38734
state_senate_district        32312
dtype: int64

## Exporting Dataframes

In [30]:
# Exporting dataframe as a pickle file
df_gun_clean.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'cleaned_gun_violence_data.pkl'))